### ISSUE TITLE
make javascript fails with SpiderMonkey 60 on Debian Buster

### ISSUE THREAD

**Issue by wohali** [apache/couchdb#2397]
@jiangphcn please see:

https://ci-couchdb.apache.org/blue/organizations/jenkins/jenkins-cm1%2FPullRequests/detail/PR-2367/44/pipeline/62

This is the new CI setup, which is proposed to use the precompiled Debian Buster libmozjs60 that the OS provides (package libmozjs60-dev).

The error we're getting is:

```
[2020-01-04T22:31:34.226Z] test/javascript/tests/view_include_docs.js                     
[2020-01-04T22:31:34.226Z]     Error: Compilation of the map function in the 'with_id' view failed: (new String("out of memory")) (function(doc) {if(doc.link_id) { var value = {'_id':doc.link_id}; if (doc.link_rev) {value._rev = doc.link_rev}; emit(doc._id, value);}};)
[2020-01-04T22:31:34.226Z] Trace back (most recent call first):
[2020-01-04T22:31:34.226Z]     
[2020-01-04T22:31:34.226Z]  551:14: test/javascript/couch.js
[2020-01-04T22:31:34.226Z]       CouchError
[2020-01-04T22:31:34.226Z]  514:12: test/javascript/couch.js
[2020-01-04T22:31:34.226Z]       CouchDB.maybeThrowError
[2020-01-04T22:31:34.226Z]  58:5: test/javascript/couch.js
[2020-01-04T22:31:34.226Z]       CouchDB/this.save
[2020-01-04T22:31:34.226Z]  41:5: test/javascript/tests/view_include_docs.js
[2020-01-04T22:31:34.226Z]       couchTests.view_include_docs
[2020-01-04T22:31:34.226Z]  45:5: test/javascript/cli_runner.js
[2020-01-04T22:31:34.226Z]       runTest
[2020-01-04T22:31:34.226Z]  56:1: test/javascript/cli_runner.js
[2020-01-04T22:31:34.226Z]       
[2020-01-04T22:31:34.226Z] fail
[2020-01-04T22:31:34.226Z] =======================================================
[2020-01-04T22:31:34.226Z] JavaScript tests complete.
[2020-01-04T22:31:34.226Z]   Failed: 1.  Skipped or passed: 76.
```

Can you have a look?  /cc @janl

**Comment by davisp** [comment]
https://gist.github.com/davisp/0e412b94456e92253f73079d0c7755f2 via [1]

I tracked down the rules file for the deb package and I'm not seeing anything that would explain a difference between this and other platforms.

Unfortunately this does not appear to be an easy fix. Looking back at the code we run a full GC right before we attempt to compile each function and then a `Maybe_GC` inbetween each function compilation.

[1] http://deb.debian.org/debian/pool/main/m/mozjs60/mozjs60_60.2.3-3.debian.tar.xz

**Comment by davisp** [comment]
Doing a bit of spelunking through the JS sources between 1.8.5 and 60 and I'm not seeing anything obviously different between the two in how we're setting stack sizes. Granted we're not rewriting source code on 1.8.5 so hard to tell there for sure.

It might be useful to try a couple tests increasing the stack size in `./dev/run` for the JS tests to see if we're borderline or if something crazy is happening.

Also I've stumbled over `export MOZ_GCTIMER=stderr` which appears to spit out some GC related information. Its a bit dense and appears to be mostly related to timing information but I am seeing a `HeapSize: 59.090 MiB` line when I run the `view_include_docs.js` test locally. I have no idea if that's close to our 64MiB limit though.

If anyone has a Buster setup to try that locally it might be useful to check what it shows. But it also makes me wonder if attempting to bump by a few more MiB would be enough to give us the breathing room there?

**Comment by davisp** [comment]
I've managed to create a somewhat duplication using a simplish approach:

Download `fail.js` from here:

https://gist.github.com/davisp/8a709a3216c1ac7e90cb1d10ccdc08c1

And then you can duplicate via:

`cat fail.js | ./src/couch/priv/couchjs share/server/main.js`

The segfault at the end isn't a big deal. We just need to scope out the AutoRequest so that it closes before we destroy the context.

**Comment by davisp** [comment]
Hurms. I'm not 100% certain whether that's a valid duplication or if its just me filling out the State.funs array.

Also a quick attempt at scoping out the request didn't work. Though given that we're exiting the process just not attempting to destroy things seems like it might be valid as well.

**Comment by davisp** [comment]
This simple diff fixes the segfault:

```diff
diff --git a/src/couch/priv/couch_js/60/main.cpp b/src/couch/priv/couch_js/60/main.cpp
index c92cfb7c9..ecedfbd3b 100644
--- a/src/couch/priv/couch_js/60/main.cpp
+++ b/src/couch/priv/couch_js/60/main.cpp
@@ -487,8 +487,5 @@ main(int argc, const char* argv[])
         JS_MaybeGC(cx);
     }
 
-    JS_DestroyContext(cx);
-    JS_ShutDown();
-
     return 0;
 }
```

**Comment by davisp** [comment]
Playing with that `fail.js` example and `MOZ_GCTIMER` does seem to show that heap line is us getting close the default 64MiB threshold. What I'm currently not understanding is how we get so close to that threshold within a single test. So it might be worth looking into what sort of garbage we might be generating that isn't going away with the multiple GC runs.

**Comment by davisp** [comment]
Last update and I've realized I'm looking at two different couchjs processes. The output for the `couchjs` that is having an error on buster appears to be closer to 1MiB. For reference I was doing this:

```bash
$ export MOZ_GCTIMER=stderr
$ make javascript suites=view_include_docs.js
```

I've just realized the output dumped to stderr of the `make javascript` terminal is from the test runner process. The GC stats for the `couchjs` that's compiling functions ends up in `dev/log/node1.log` and those numbers are much smaller around 1MiB instead of the nearly 64MiB I was seeing for the test runner process.

That's all I've got for the evening. If anyone has a better idea on what to try tracking down tomorrow I'm all ears.

**Comment by jiangphcn** [comment]
thanks @davisp and @wohali for this. We can see `out of memory"` for one test case. I will continue to investigate based on above from Paul.

**Comment by jiangphcn** [comment]
hey @davisp , @wohali  , @janl , after some investigation, I highly suspect that this `"out of memory"` is related to `infinite_loop` test in https://github.com/apache/couchdb/blob/master/test/javascript/tests/view_errors.js#L154
```
      var designDoc3 = {
        _id:"_design/infinite",
        language: "javascript",
        views: {
          "infinite_loop" :{map:"function(doc) {while(true){emit(doc,doc);}};"}
        }
      };
      T(db.save(designDoc3).ok);

      try {
          db.view("infinite/infinite_loop");
          T(0 == 1);
      } catch(e) {
          T(e.error == "os_process_error" || e.error == "unnamed_error");
      }
```

The sequence of test suites follows `view_errors.js` and then `view_include_docs.js`. In some case, `couchjs` is shared among different test suites. After `view_errors` bumped memory, there is no room for `view_include_docs.js` to allocate when the shared `couchjs` is used.

Today, I ever made several test in docker image
1. run `make javascript suites="view_include_docs"` more than 200 times in the loop, and didn't hit out-of-memory issue
2. run `make javascript suites="view_errors  view_include_docs"`, and i can get out-of-memory sometime, like

```
root@0010847bf784:/couchdb# make javascript suites="view_errors  view_include_docs"
test/javascript/tests/view_errors.js         pass
test/javascript/tests/view_include_docs.js
    Error: (new String("out of memory")) (function (doc) { if (typeof doc.integer === 'number') {emit(doc.integer, doc.integer)};})
Trace back (most recent call first):

 551:14: test/javascript/couch.js
      CouchError
 514:12: test/javascript/couch.js
      CouchDB.maybeThrowError
 202:5: test/javascript/couch.js
      CouchDB/this.view
 43:14: test/javascript/tests/view_include_docs.js
      couchTests.view_include_docs
 45:5: test/javascript/cli_runner.js
      runTest
 56:1: test/javascript/cli_runner.js

fail
=======================================================
JavaScript tests complete.
  Failed: 1.  Skipped or passed: 1.
make: *** [Makefile:279: javascript] Error 1
```

3. run `make javascript suites="view_errors  view_include_docs"` after commenting out below in `view_errors.js`. I didn't see out-of-memory when running more than 100 times 
```
      try {
          db.view("infinite/infinite_loop");
          T(0 == 1);
      } catch(e) {
          //T(e.error == "os_process_error" || e.error == "unnamed_error");
      }
```

Could we give me a try in Jenkins server to disable infinite_loop temporarily?

**Comment by janl** [comment]
@jiangphcn good find! I wonder what happens if we add a `gc();` call in our loop.js that just cleans things up periodically.

FTR, I can’t get this to fail with just the two tests on my Mac setup here.

**Comment by janl** [comment]
@davisp that patch def helps segfaults here as well. funnily enough they don’t seem to be hit when running our tests and whatnot, but running custom code in couches fails quickly. I’ll prepare a PR for this.

**Comment by jiangphcn** [comment]
> FTR, I can’t get this to fail with just the two tests on my Mac setup here.

same for me. I didn't see this issue in my Mac environment. Also, it is not always happening in Linux environment. I setup environment today using docker image and can see that this is randomly happening. 

> I wonder what happens if we add a gc(); call in our loop.js that just cleans things up periodically.

I can have a try to add gc(); to see whether it helps.

PS: for jenkins test, I didn't figure out how to login to trigger rebuild so this is why I asked before to give a try with disable infinite_loop temporarily?

**Comment by davisp** [comment]
@janl Its just when couchjs exits and we haven't properly terminated the autorequest thing. Its a bit of a silly assertion during cleanup that we don't care about since we're exiting. It had me more concerned until I ran it through the debugger.

**Comment by davisp** [comment]
This will be fixed when #2409 is merged.
